Purchases/Caching/RCDeviceCache.m

```diff
@@ -284,35 +287,77 @@ - (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
         NSMutableDictionary <NSString *, NSDictionary *>
             *groupedAttributes = self.storedAttributesForAllUsers.mutableCopy;
         [groupedAttributes removeObjectForKey:appUserID];
-        [self.userDefaults setObject:groupedAttributes forKey:RCSubscriberAttributesKey];
+        [self storeAttributesForAllUsers:groupedAttributes];
     }
 }
 
 # pragma mark - subscriber attributes migration from per-user key to grouped key
 
-- (void)migrateSubscriberAttributesIfNeededForAppUserID:(nonnull NSString *)appUserID {
+- (void)cleanupSubscriberAttributes {
     @synchronized (self) {
-        NSDictionary *legacySubscriberAttributes = [self valueForLegacySubscriberAttributes:appUserID];
-        if (legacySubscriberAttributes != nil) {
-            [self migrateSubscriberAttributes:legacySubscriberAttributes withAppUserID:appUserID];
-        }
+        [self migrateSubscriberAttributes];
+        [self deleteSyncedSubscriberAttributesForOtherUsers];
     }
 }
 
-- (void)migrateSubscriberAttributes:(nonnull NSDictionary *)legacyAttributes
-                      withAppUserID:(nonnull NSString *)appUserID {
-    NSMutableDictionary *allSubscriberAttributes = ([self.userDefaults dictionaryForKey:RCSubscriberAttributesKey]
-                                                    ?: @{}).mutableCopy;
-    NSMutableDictionary *currentAttributesForAppUserID = allSubscriberAttributes[appUserID]
-                                                         ?: @{}.mutableCopy;
-    NSMutableDictionary *mutableLegacyAttributes = legacyAttributes.mutableCopy;
-    [mutableLegacyAttributes addEntriesFromDictionary:currentAttributesForAppUserID];
+- (void)migrateSubscriberAttributes {
+    NSArray *appUserIDsWithLegacyAttributes = [self appUserIDsWithLegacyAttributes];
+    NSMutableDictionary *attributesInNewFormat = self.storedAttributesForAllUsers.mutableCopy;
+    for (NSString *appUserID in appUserIDsWithLegacyAttributes) {
+        NSDictionary *legacyAttributes = ([self valueForLegacySubscriberAttributes:appUserID] ?: @{}).mutableCopy;
+        NSMutableDictionary *currentAttributesForAppUserID = (legacyAttributes ?: @{}).mutableCopy;
+        NSMutableDictionary *allAttributesForUser = legacyAttributes.mutableCopy;
+        [allAttributesForUser addEntriesFromDictionary:currentAttributesForAppUserID];
```

Comment:
I might be missing something here but:

If `currentAttributesForAppUserID` is a copy of `legacyAttributes` and  `allAttributesForUser` is also a copy of `legacyAttributes`, isn't this line redundant since `allAttributesForUser` is already equal to `currentAttributesForAppUserID`?

If this line is redundant, then `currentAttributesForAppUserID` and `allAttributesForUser` are the same, and therefore all this can be simplified to:

```
        NSDictionary *legacyAttributes = [self valueForLegacySubscriberAttributes:appUserID] ?: @{};
        attributesInNewFormat[appUserID] = legacyAttributes;
```

Reply:
yep. I got dizzy while writing this. I'll clean it up, thanks for catching

Reply:
just realized what the mixup was here: 
I meant to make sure that when migrating, if there happen to be any attributes in the new format already, those are kept. 
so line 309 should have been 
```objective-c
NSMutableDictionary *currentAttributesForAppUserID = (self.storedAttributesForAllUsers ?: @{}).mutableCopy;
```
I'll update, try to clean up for readability and add a test for the specific case

Reply:
although it should never happen, but still

Reply:
> I meant to make sure that when migrating, if there happen to be any attributes in the new format already, those are kept.

Ok perfect then!